src/supervisor/Supervisor.cc

```diff
@@ -167,20 +173,83 @@ static std::vector<std::string> extract_msgs(std::string* buffer, char delim)
 	return rval;
 	}
 
-static std::pair<int, std::vector<std::string>> read_msgs(int fd, std::string* buffer, char delim)
+// Read some bytes from the given fd and append them to buffer.
+static int read_bytes_into_buffer(int fd, std::string* buffer)
 	{
 	constexpr auto buf_size = 256;
 	char buf[buf_size];
 
 	int bytes_read = read(fd, buf, buf_size);
 
 	if ( bytes_read <= 0 )
-		return {bytes_read, {}};
+		return bytes_read;
 
 	buffer->append(buf, bytes_read);
+
+	return bytes_read;
+	}
+
+static std::pair<int, std::vector<std::string>> read_msgs(int fd, std::string* buffer, char delim)
+	{
+	int bytes_read = read_bytes_into_buffer(fd, buffer);
+	if ( bytes_read < 0 && errno == EAGAIN ) // Treat EAGAIN as no bytes read.
+		bytes_read = 0;
+
+	if ( bytes_read < 0 )
+		return {bytes_read, {}};
+
 	return {bytes_read, extract_msgs(buffer, delim)};
 	}
 
+// Write a message to the OutFD() of the given pipe pair.
+//
+// Should the write() fail with EAGAIN, usleep() for a tiny amount
+// and retry after draining data pending on InFD() and appending
+// it to given buffer. Draining is done in case the other side of the
+// pipe is also blocking on a write() that we're blocking.
+static bool write_msg(const std::unique_ptr<detail::PipePair>& pipe, const std::string& msg,
+                      std::string* buffer)
+	{
+	// Send the string including its '\0'
+	int len = msg.size() + 1;
+	const char* data = msg.data();
+
+	while ( len > 0 )
+		{
+		int n = write(pipe->OutFD(), data, len);
+
+		if ( n < 0 )
+			{
+			if ( errno == EINTR )
+				continue;
+
+			// If there's no room in the pipe and we would've blocked, relax
+			// a bit, but otherwise just retry again. No fancy polling.
+			//
+			// We drain the stem's queue for a best effort preventing it
+			// running into a full pipe as well.
+			if ( errno == EAGAIN || errno == EWOULDBLOCK )
```

Comment:
There was a tiny bit worry that Supervisor/stem deadlock each other (though I doubt it would be an actual issue in practice) which is why I added the reading from the other end pipe.

I'm not very much liking the new approach, but also contrasting it with the `abort()` that's happening today.

FWIW: The other idea was to make the pipe blocking instead of the usleep/spinning here, but you'd still need to do something to avoid the potential deadlocks.